First cut at new counter handling.

pilhuhn opened the pull request:
Introduces ConcurrentGauge and removed monotonic flag from Counted
annotation, changing its default behaviour.

Signed-off-by: Heiko W. Rupp <hrupp@redhat.com>

OndrejM commented on spec/src/main/asciidoc/app-programming-model.adoc:
```diff
@@ -190,24 +191,15 @@ The implementation must support the following annotation targets:
 
 When a constructor is annotated, the implementation must register a counter for the constructor using the <<annotated-naming-convention>>.
-The counter is treated according to the `monotonic` field when the constructor is invoked.
+The counter is is increased by one when the constructor is invoked.
```
Grammar: "is" is doubled

OndrejM commented on spec/src/main/asciidoc/architecture.adoc:
```diff
@@ -127,7 +127,8 @@ The Metadata:
 * unit: a fixed set of string units
 * type:
-** counter: an incrementally increasing or decreasing numeric value (e.g. total number of requests received or total number of concurrently active HTTP sessions).
+** counter: an monotonically increasing numeric value (e.g. total number of requests received).
```
Grammar: Should be "a monotonically"

pilhuhn commented:
Thanks @OndrejM Yes, we need to add a few things to the change log - not only this :)

jmartisk commented:
I'm missing additions related to `ConcurrentGauge` in `org.eclipse.microprofile.metrics.MetricRegistry`.

Emily-Jiang commented on api/src/main/java/org/eclipse/microprofile/metrics/annotation/ConcurrentGauge.java:
```diff
@@ -0,0 +1,138 @@
+@Documented
+@InterceptorBinding
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE, ElementType.CONSTRUCTOR, ElementType.METHOD, ElementType.ANNOTATION_TYPE })
```
I am not sure why ElementType.ANNOTATION_TYPE should be included. I cannot find a use case for such an annotation class to have this binding applied.

pilhuhn replied:
TBH I have no idea :-)  I copied it from `Counted`, but I guess it does not make sense there either. But then what do I know :)

OndrejM replied:
Having annotations on other annotations makes sense for CDI stereotypes. However, I'm not sure the stereotypes are applicable to interceptor bindings, maybe @antoinesd knows?

OndrejM replied:
Apparently according to the Java EE tutorial, it makes sense to add interceptors to stereotype annotations: https://docs.oracle.com/javaee/6/tutorial/doc/gkhqc.html
So having `ElementType.ANNOTATION_TYPE` here really makes sense.

Emily-Jiang replied:
Not really. You see ApplicationScoped has the target of @Target({ TYPE, METHOD, FIELD }). It can be part of stereotype without any problem. Type means class, which includes annotation type. Therefore, I think we should not include the ANNOTATION_TYPE.

OndrejM replied:
You're right, @Emily-Jiang, thanks for correcting me. I studied a bit more and ANNOTATION_TYPE seems to be a subset of TYPE. It makes sense only if the usage should be restricted to annotations (for meta annotations) and not other types. Having TYPE already allows using an annotation with other annotations and it's possible to add them to a stereotype.

We can remove ANNOTATION_TYPE as @Emily-Jiang suggested.

Emily-Jiang commented on api/src/main/java/org/eclipse/microprofile/metrics/ConcurrentGauge.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * A concurrent gauge is a gauge that measures parallel invocations of a method.
+ *
+ * In MicroProfile-Metrics before 2.0 this was done via @Counted(monotonic=false),
```
I don't think this should be mentioned. It is better to mentioned in the spec instead of the javadoc in the release note. I think this clatters the javadoc.

OndrejM replied:
I agree, it should be mentioned (and already is) in the spec in the release notes and appendix about upgrading from the previous version.

Emily-Jiang commented on api/src/main/java/org/eclipse/microprofile/metrics/ConcurrentGauge.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public interface ConcurrentGauge extends Metric {
+
+  long getCount();
```
It is better to have a sentence javadoc.

jmartisk commented on spec/src/main/asciidoc/rest-endpoints.adoc:
```diff
@@ -390,6 +405,34 @@ Counters do not have a suffix for the unit.
+# TYPE application_visitors_total counter
+# HELP application_visitors_total The number of parallel invocations of visitorsTotal()
+application_visitors_total 80
+application_visitors_min 20
```
The `min` and `max` values should probably have a TYPE line, with the type being... gauge?

pilhuhn replied:
Yes, this is an error on my side. The original proposal had `_current`

pilhuhn replied:
Actually I wonder if those should be 3 metrics or 1 metric with 3 different labels
```
method_a_invocations{???="min"}
method_a_invocations{???="max"}
method_a_invocations{???="current"}

```
( I obviously can't come up with a good name for the key here)

donbourne replied:
I would suggest it be 3 different metrics.  When you use labels, the thing you're labeling should make sense to sum or average across all sets of labels -- which wouldn't be the case here.  See https://prometheus.io/docs/practices/naming/ ... 

> As a rule of thumb, either the sum() or the avg() over all dimensions of a given metric should be meaningful (though not necessarily useful). If it is not meaningful, split the data up into multiple metrics. For example, having the capacity of various queues in one metric is good, while mixing the capacity of a queue with the current number of elements in the queue is not.

donbourne commented on tck/api/src/main/java/io/astefanutti/metrics/cdi/se/ConcurrentGaugedConstructorBean.java:
```diff
@@ -15,11 +15,11 @@
+import org.eclipse.microprofile.metrics.annotation.ConcurrentGauge;
 
-public class CountedConstructorBean {
+public class ConcurrentGaugedConstructorBean {
```
Looks like this class (and a number of others) was renamed -- we should keep the old Counter tests working as before

pilhuhn replied:
Well, the old counters are gone, so I've moved them to Monotonic*Test.
So they are not lost.

donbourne commented on tck/api/src/main/java/io/astefanutti/metrics/cdi/se/ConcurrentGaugedClassBeanTest.java:
```diff
@@ -39,20 +62,20 @@
 
 @RunWith(Arquillian.class)
-public class CountedClassBeanTest {
+public class ConcurrentGaugedClassBeanTest {
```
This class is still testing Counters (not gauges)

jmartisk commented on tck/api/src/main/java/io/astefanutti/metrics/cdi/se/ConcurrentGaugedMethodBeanTest.java:
```diff
@@ -64,33 +64,33 @@
-    public void countedMethodNotCalledYet(@Metric(name = "countedMethod", absolute = true) Counter instance) {
-        assertThat("Counter is not registered correctly", registry.getCounters(), hasKey(COUNTER_NAME));
-        Counter counter = registry.getCounters().get(COUNTER_NAME);
+    public void countedMethodNotCalledYet(@Metric(name = "countedMethod", absolute = true) ConcurrentGauge instance) {
```
The metric name here should be `C_GAUGE_NAME`?

Channyboy replied:
right, it should be changed to `C_GAUGE_NAME`

jmartisk commented on tck/api/src/main/java/io/astefanutti/metrics/cdi/se/ConcurrentGaugedClassBeanTest.java:
```diff
@@ -84,6 +108,11 @@ public void callCountedMethodsOnce() {
+
+        // Make sure that the max is one
+        assertThat("Concurrent Gauges counts are incorrect", registry.getConcurrentGauges().values(),
+                   everyItem(Matchers.<ConcurrentGauge>hasProperty("count", equalTo(1L))));
```
Here the comment is talking about `max` but assertion is checking `count`

jmartisk commented:
I see some assertions about the `max` value failing because `max` should, as far as I understand, contain the value valid for the previous *full minute*, but the tests don't wait until the next minute starts and rather check the `max` right after invocations finish, so it contains 0 (unless the execution was timed 'just right' so that a new minute starts at the correct moment).

This is quite a tricky thing to test - we need to make sure our desired invocations are performed within one complete minute (so it's a bit dangerous to start a test at a time ending with :59, for example), and then to verify the results, the test has to wait until the start of the next full minute, so the `min` and `max` numbers are updated. Another downside is that such test will obviously take at least a minute to complete.

I have an implementation of a functional test for `min` and `max` here: https://github.com/jmartisk/microprofile-metrics/commits/conc-gauge-test Given that this test takes around two minutes, I would suggest using this one and not dealing very much with `min` and `max` values in other tests?

Channyboy commented on api/src/main/java/org/eclipse/microprofile/metrics/annotation/ConcurrentGauge.java:
```diff
@@ -0,0 +1,128 @@
+    @Nonbinding
+    String description() default "";
+
+
```
Should there be a unit's parameter to match up with the other annotations?

pilhuhn replied:
Theoretically yes to be consistent with other metrics. In reality the unit would always be NONE.

pilhuhn commented:
I think I have incorporated most (all?) feedback and rebased to current master (after the tags changes)
I  propose looking at this and if there are no major issues, to merge and then refine from there.

Channyboy commented on spec/src/main/asciidoc/rest-endpoints.adoc:
```diff
@@ -451,6 +466,38 @@ Counters do not have a suffix for the unit.
+.Example OpenMetrics text format for a Concurrent Gauge
+[source, ruby]
+----
+# TYPE application_method_a_invocations_current gauge
```
concurrentGauge instead of gauge?

jmartisk replied:
No, this is `gauge` as an OpenMetrics type (https://prometheus.io/docs/concepts/metric_types/#gauge), OpenMetrics doesn't have a `concurrentgauge` type

pilhuhn commented:
@donbourne No. See 
https://www.eclipse.org/projects/handbook/#ip-copyright-headers

And from the JakartaEE ML: 

> The short version is that we expect to see the year that the content was initially created. Project teams can optionally decide to include a "most recently changed" year as well.
